### ISSUE TITLE
Enforce checks on input spreadsheet structure

### ISSUE THREAD

**Issue by ollylucl** [CLOSER-Cohorts/archivist-utilities#50]
In order to perform validation and error-checking on files uploaded to the tool, the application needs to enforce the following structure on the input files. If an input file does not have the structure defined below (i.e. worksheets and columns with particular names) no output files are generated and an error message is displayed to the user.

The input files MUST:
- contain a worksheet titled 'QV and TV Mappings'
- contain a worksheet titled 'DV'

The titles of these worksheets are not case-sensitive, i.e. 'qv and tv Mappings', 'QV AND TV MAPPINGS', 'dv', 'Dv', etc. are all valid titles.

The position of these worksheets in the sequence of available worksheets is not important; the application code should select worksheets by their title, not by their position in the spreadsheet.

The 'QV and TV Mappings' worksheet MUST contain column headings with the following titles:

- Questionnaire prefix
- Question name
- Dataset prefix
- Variable name
- Topic id

The 'DV' worksheet MUST contain column headings with the following titles:

- Derived Dataset prefix
- Derived Variable Name
- Source Dataset prefix
- Source Variable Name

Although these column names must be defined in the 'QV and TV Mappings' and 'DV' worksheets, there do not have to be row values specified for these columns. With the current implementation of the application, if there is insufficient valid data to generate the content for an output file (i.e. a row in that file, or if there are no valid rows the file itself) that output is not generated.

The column headings are not case-sensitive (e.g. 'derived dataset prefix', 'DERIVED DATASET PREFIX', 'TOPIC id', 'Question Name', etc. are all valid column names). The columns do not have to appear in a particular sequence in the worksheet, the application code selects columns by their name, not their position in a worksheet.  

The application code needs to be updated to use these new rules around the structure of input spreadsheets; the code needs to select worksheets/columns by name, not by their position. The README file needs to be updated to reflect these changes as well.

**Comment by ollylucl** [comment]
Also, is it ok to call column A in the DV worksheet 'Primary dataset prefix' and column D 'Secondary dataset prefix'? We need to give them different names, are those names ok?

**Comment by ollylucl** [comment]
When creating the *_dv.txt files, we create a separate .txt file for each unique value in the dataset_prefix column in the DV worksheet. Which dataset_prefix column should we use, column A or column B?

**Comment by HayleyMills** [comment]
We're using the template (mappings_wip_template.xlsx) on this page, I'll update the wiki. https://wiki.ucl.ac.uk/display/CTTEAM/Mappings

**Comment by ollylucl** [comment]
Ok, thanks @HayleyMills. Part of this issue involves renaming the dataset prefix columns in the DV worksheet so they have unique names. Are you ok with the names that I've suggested? If so shall I update the mappings_wip_template.xlsx on https://wiki.ucl.ac.uk/display/CTTEAM/Mappings?

**Comment by HayleyMills** [comment]
@ollylucl apologies, I updated the spreadsheet after our last meeting in the office, but didn't replace it in the wiki! sorry. It should be okay now, let me know if not.

**Comment by ollylucl** [comment]
Ah ok, thanks for that, I have the updated template now. I'l update this issue so it uses the column names in the new version.

**Comment by ollylucl** [comment]
```When creating the *_dv.txt files, we create a separate .txt file for each unique value in the dataset_prefix column in the DV worksheet.```

Just a reminder about this query, @HayleyMills . Which column in the 'DV' worksheet do we use for generating filenames. 'Derived Dataset prefix' or 'Source Dataset prefix'?

**Comment by HayleyMills** [comment]
@ollylucl good question, most of the time they will be the same, but I would use the first one which is the Derived.

**Comment by ollylucl** [comment]
I've pushed up a commit to github which makes the changes to the functionality of the mappings generator as described in this issue. The README file has also been updated to reflect these changes. https://closer-cohorts.github.io/archivist-utilities/ has been updated so it has the functionality described in this issue (i.e. it's selecting worksheets based on their title, not their position in the spreadsheet). I've added a little more error checking/reporting. Hopefully this will make the application a bit more robust.

The exact changes I've made to the code and README are in these pull requests:

https://github.com/CLOSER-Cohorts/archivist-utilities/pull/51
https://github.com/CLOSER-Cohorts/archivist-utilities/pull/52
